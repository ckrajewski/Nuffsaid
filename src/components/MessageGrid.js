import React, { useState, useEffect } from 'react';
import Grid from '@material-ui/core/Grid';
import Column from './Column';

/**
* MessageGrid component renders the column grids
* Error, Warning, and Info (depending on values passed in by the
* prioirities prop)
*/
export default function MessageGrid(props) {
  const { message, priorities, setClear } = props;
  const { priority } = message;

  /**
  * defaults filteredMessages to an object, where each key
  * is a given priority type. This is also how each column
  * is rendered (each column is rendered by prioirty type)
  */
  const initializeFilteredMessages = () => (
    priorities.reduce((filteredMessages, priority) => {
      filteredMessages[priority.value] = [];
      return filteredMessages;
    }, {}));

  /**
  * filtered messages is a state variable responsible for storing our
  * messages that are already filtered by error type
  */
  const [filteredMessages, setFilteredMessages] = useState(() => initializeFilteredMessages());

  /**
  * everytime a new message comes in, add it
  * to the appropriate column (priority type)
  */
  useEffect(() => {
    setFilteredMessages({
      ...filteredMessages,
      [priority]: [message, ...filteredMessages[priority].slice()],
    });
  }, [message]);

  /**
  * clear all messages by re-initializing the state of filteredMessages object
  */
  const handleClearAll = () => setFilteredMessages(initializeFilteredMessages());

  /**
  * intialize higher order function setClear
  * to take handleClearAll as its argument
  */
  useEffect(() => {
    setClear(handleClearAll);
  }, []);

  /**
  * removes given message from the array by its index
  */
  const handleClearMessage = (priority, index) => {
    const filteredMessagesByPriorityCopy = [...filteredMessages[priority]];
    filteredMessagesByPriorityCopy.splice(index, 1);
    setFilteredMessages({
      ...filteredMessages,
      [priority]: filteredMessagesByPriorityCopy,
    });
  };

  return (
    <div>
      <Grid container spacing={2} justify="center">
        {
            priorities.map((priority) => (
              <Grid item xs={Math.floor(12 / priorities.length)} key={priority.value}>
                <Column
                  clearMessage={handleClearMessage}
                  priority={priority}
                  messages={filteredMessages[priority.value]}
                  title={priority.label}
                />
              </Grid>

            ))
        }
      </Grid>
    </div>
  );
}
