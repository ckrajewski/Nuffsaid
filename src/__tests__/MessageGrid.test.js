import React from 'react';
import { shallow, mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import MessageGrid from '../components/MessageGrid';
import renderer from 'react-test-renderer';

configure({ adapter: new Adapter() });
const message ={priority: 1, messages:'error message'};
const priorities =[{value:1,label:'Error'},{value:2,label:'Warning'}];
const setClear = jest.fn();
describe('MessageGrid component', () => {
  test('renders with default values', () => {
  	const tree = renderer.
    create(<MessageGrid message={message} priorities={priorities} setClear={setClear}/>)
    .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('test new incoming message', () => {
    const wrapper = shallow(<MessageGrid message={message} priorities={priorities} setClear={setClear} />);
    const updatedMessage ={priority: 2, messages:'warning message'};
    wrapper.setProps({message:updatedMessage});
    const tree = renderer.create(wrapper).toJSON();
    expect(tree).toMatchSnapshot();
  });
});