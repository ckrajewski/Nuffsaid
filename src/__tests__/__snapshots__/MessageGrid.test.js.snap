// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageGrid component renders with default values 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-justify-xs-center"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div>
        <div>
          <div>
            <h6
              className="MuiTypography-root MuiTypography-h6"
            >
              Error
               
              Type
               
              1
            </h6>
          </div>
          <div>
            <p
              className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
            >
              Count
               
              0
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div>
        <div>
          <div>
            <h6
              className="MuiTypography-root MuiTypography-h6"
            >
              Warning
               
              Type
               
              2
            </h6>
          </div>
          <div>
            <p
              className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
            >
              Count
               
              0
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MessageGrid component test new incoming message 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-justify-xs-center"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div>
        <div>
          <div>
            <h6
              className="MuiTypography-root MuiTypography-h6"
            >
              Error
               
              Type
               
              1
            </h6>
          </div>
          <div>
            <p
              className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
            >
              Count
               
              0
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div>
        <div>
          <div>
            <h6
              className="MuiTypography-root MuiTypography-h6"
            >
              Warning
               
              Type
               
              2
            </h6>
          </div>
          <div>
            <p
              className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
            >
              Count
               
              0
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MessageGrid component test new message 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-justify-xs-center"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div>
        <div>
          <div>
            <h6
              className="MuiTypography-root MuiTypography-h6"
            >
              Error
               
              Type
               
              1
            </h6>
          </div>
          <div>
            <p
              className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
            >
              Count
               
              0
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div>
        <div>
          <div>
            <h6
              className="MuiTypography-root MuiTypography-h6"
            >
              Warning
               
              Type
               
              2
            </h6>
          </div>
          <div>
            <p
              className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
            >
              Count
               
              0
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
